packages/cli/src/ceramic-cli-utils.ts

```diff
@@ -106,42 +100,74 @@ export class CeramicCliUtils {
         _corsAllowedOrigins = corsAllowedOrigins.split(' ').map((origin) => new RegExp(origin))
       }
       const logLevel = verbose ? LogLevel.verbose : debug ? LogLevel.debug : LogLevel.important
-      const loggerConfig: LoggerConfig = {
-        logToFiles,
-        logDirectory,
-        logLevel,
-      }
-
       const _syncOverride = SYNC_OPTIONS_MAP[syncOverride]
 
+      if (stateStoreDirectory && stateStoreS3Bucket) {
+        throw new Error(
+          'Cannot specify both --state-store-directory and --state-store-s3-bucket. Only one state store - either on local storage or on S3 - can be used at a time'
+        )
+      }
+      const stateStoreMode = stateStoreS3Bucket ? StateStoreMode.S3 : StateStoreMode.FS
+      const ipfsMode = ipfsApi ? IpfsMode.REMOTE : IpfsMode.BUNDLED
+
       configFromCli = {
-        ethereumRpcUrl: ethereumRpc,
-        anchorServiceUrl: anchorServiceApi,
-        stateStoreDirectory,
-        s3StateStoreBucket: stateStoreS3Bucket,
-        validateStreams,
-        ipfsPinningEndpoints,
-        gateway,
-        port,
-        hostname,
-        loggerConfig,
-        network,
-        pubsubTopic,
-        corsAllowedOrigins: _corsAllowedOrigins,
-        ipfsHost: ipfsApi,
-        syncOverride: _syncOverride,
+        anchor: {
+          anchorServiceUrl: anchorServiceApi,
+          ethereumRpcUrl: ethereumRpc,
+        },
+        httpApi: {
+          corsAllowedOrigins: _corsAllowedOrigins,
+          hostname,
+          port,
+        },
+        ipfs: {
+          mode: ipfsMode,
+          host: ipfsApi,
+          pinningEndpoints: ipfsPinningEndpoints,
+        },
+        logger: {
+          logDirectory,
+          logLevel,
+          logToFiles,
+        },
+        network: {
+          name: network,
+          pubsubTopic,
+        },
+        node: {
+          gateway,
+          syncOverride: _syncOverride,
+          validateStreams,
+        },
+        stateStore: {
+          mode: stateStoreMode,
+          localDirectory: stateStoreDirectory,
+          s3Bucket: stateStoreS3Bucket,
+        },
       }
 
       // Filter out undefined flags
-      configFromCli = Object.fromEntries(
-        Object.entries(configFromCli).filter(([k, v]) => v !== undefined)
-      )
+      configFromCli = this.removeUndefinedFields(configFromCli)
     }
 
     const config = Object.assign(configFromFile, configFromCli)
     return CeramicDaemon.create(config)
   }
 
+  static removeUndefinedFields(obj: Record<string, any>): Record<string, any> {
```

Comment:
I need to go on record that I hate javascript for making me write this.

@ukstv, did I miss some built-in function that would have done this for me?